File: Data Structures & Algorithms/find-minimum-in-rotated-sorted-array/submission-0.py
class Solution:
    def findMin(self, nums: List[int]) -> int:
        l, r = 0, len(nums) - 1
        while r-l > 1:
            mid = (l + r) // 2
            print(nums[l],nums[mid], nums[r],l,mid,  r)
            if nums[mid] < nums[r] and nums[mid] < nums[l]:
                r = mid
            elif nums[mid] < nums[r] and nums[mid] > nums[l]:
                return nums[l]
            elif nums[mid] > nums[r]:
                l = mid + 1
        return nums[l] if nums[l] < nums[r] else nums[r]

File: Data Structures & Algorithms/combination-target-sum/submission-0.cpp
class Solution {
public:
    vector<vector<int>> combinationSum(vector<int>& nums, int target) {
        vector<int> subset;
        vector<vector<int>> res;
        dfs(nums,subset,target,0,res);
        return res;
    }
private:
    void dfs(vector<int>& nums,vector<int>& subset,int target, int i, vector<vector<int>>& res) {
        if (target == 0) {
            res.push_back(subset);
            return;
        }
        if (target < 0) {
            return;
        }
        if (i == nums.size()){
            return;
        }
        // TAKE nums[i]
        subset.push_back(nums[i]);
        dfs(nums, subset, target - nums[i], i,res);   // stay at i (reuse allowed)
        subset.pop_back();

        // SKIP nums[i]
        dfs(nums, subset, target, i + 1,res);
    }

};
